### PR TITLE
feature(tracing): Add React Native Gesture Handler tracing

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description

[RNGH](https://github.com/software-mansion/react-native-gesture-handler/) is a very popular gesture-handling library for RN. This PR adds a wrapper that makes it simple to trace the gesture interactions.

## :bulb: Motivation and Context

- https://github.com/getsentry/sentry-react-native/pull/2835
- closes: https://github.com/getsentry/sentry-react-native/issues/2753

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [x] All tests passing
- [x] No breaking changes

## :crystal_ball: Next steps
- [x] waits for https://github.com/getsentry/sentry-react-native/pull/2835 and JS release

### CODE REVIEW

File: src/js/tracing/ops.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ export const NAVIGATION = 'navigation';
 
 export const UI_LOAD = 'ui.load';
 export const UI_ACTION_TOUCH = 'ui.action.touch';
+export const UI_ACTION_GESTURE = 'ui.action.gesture';
```
Comment (marandaneto) [inline src/js/tracing/ops.ts:7]: What type of gestures are there?

Comment (marandaneto) [inline src/js/tracing/ops.ts:7]: Can we build it dynamically? eg `ui.action.*` where * is the gesture name.

File: src/js/tracing/gesturetracing.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+import { ACTION_GESTURE_OP } from './operations';
+import { ReactNativeTracing } from './reactnativetracing';
+
+export const DEFAULT_BREADCRUMB_CATEGORY = 'gesture';
+export const DEFAULT_BREADCRUMB_TYPE = 'user';
+
+/**
+ * Internal interface following RNGH 2 Gesture Event API.
+ * We need to use this to avoid importing RNGH 2 types and depending on it.
+ * https://github.com/software-mansion/react-native-gesture-handler/blob/2.9.0/src/handlers/gestures/gesture.ts#L55
```
Comment (marandaneto) [inline src/js/tracing/gesturetracing.ts:14]:
```suggestion
 * https://github.com/software-mansion/react-native-gesture-handler/blob/f0868f7ccf678c947ef65519ebf97ae149a10289/src/handlers/gestures/gesture.ts#L55
```

Always prefer permalinks

File: src/js/tracing/ops.ts (hunk 1 above)
Comment (krystofwoldrich) [reply src/js/tracing/ops.ts:7]: Not that many.

```
Fling
ForceTouch
LongPress
NativeView
Pan
Pinch
Rotation
Tap
Manual
```

File: src/js/tracing/gesturetracing.ts (hunk 2 above)
Comment (krystofwoldrich) [reply src/js/tracing/gesturetracing.ts:14]: Ok, thanks.

File: src/js/tracing/ops.ts (hunk 1 above)
Comment (krystofwoldrich) [reply src/js/tracing/ops.ts]: Yes.

**Comment by krystofwoldrich** [comment]
@marandaneto Can this get a final review? 🙏

File: sample-new-architecture/src/Screens/GesturesTracingScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    .onBegin(() => {
+      console.log('Pinch gesture begin');
+    })
+    .onEnd(() => {
+      console.log('Pinch gesture end');
+      startExampleSpan();
+    });
+
+  return (
+    <GestureDetector gesture={traceGesture('pinch', gesture)}>
```
Comment (marandaneto) [inline sample-new-architecture/src/Screens/GesturesTracingScreen.tsx:17]: Is this the wrapper that people need to do in order to trace the gestures?
M: Maybe we call it `sentryTraceGesture or something like that? right now it's unclear if this is related to sentry instrumentation or not.

File: sample-new-architecture/src/Screens/GesturesTracingScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import { getCurrentHub, Scope, traceGesture } from '@sentry/react-native';
+
+const GesturesTracingScreen = () => {
+  const gesture = Gesture.Pinch()
+    .onBegin(() => {
+      console.log('Pinch gesture begin');
+    })
+    .onEnd(() => {
+      console.log('Pinch gesture end');
+      startExampleSpan();
```
Comment (marandaneto) [inline sample-new-architecture/src/Screens/GesturesTracingScreen.tsx:13]: Should the example start during `onBegin` and end during `onEnd`? its confusing right now.

File: src/js/tracing/gesturetracing.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+   * Label of the gesture to be used in transaction name.
+   * Example: dismissGesture
+   */
+  label: string,
+  gesture: GestureT,
+  options: Partial<GestureTracingOptions> = {},
+): GestureT {
+  const gestureCandidate = gesture as unknown as BaseGesture | undefined | null;
+  if (!gestureCandidate) {
+    logger.warn('[ReactNativeTracing] Gesture can not be undefined');
```
Comment (marandaneto) [inline src/js/tracing/gesturetracing.ts:55]: L: Would make sense to use `[GestureTracing]` instead? so you know where it's coming from, not sure if the rest of the SDK does that.

**Review by marandaneto — APPROVED** [review]
Left a few comments, but other than that LGTM

File: sample-new-architecture/src/Screens/GesturesTracingScreen.tsx (hunk 3 above)
Comment (krystofwoldrich) [reply sample-new-architecture/src/Screens/GesturesTracingScreen.tsx:17]: I see, yes, that makes sense to me.

File: sample-new-architecture/src/Screens/GesturesTracingScreen.tsx (hunk 4 above)
Comment (krystofwoldrich) [reply sample-new-architecture/src/Screens/GesturesTracingScreen.tsx:13]: I wanted to showcase that you can start a fetch or something and the Sentry will trace it.

So the end of the gesture triggers it and then you can see in Sentry what happened after that.

Maybe I should rename the `startExampleSpan` function to be more clear to something like `fetchExampleData`.

File: src/js/tracing/gesturetracing.ts (hunk 5 above)
Comment (krystofwoldrich) [reply src/js/tracing/gesturetracing.ts:55]: At least the `[ReactNativeTracing]` does it. Makes sense. Thanks.

**Pull request merged**